### PR TITLE
(0.90.0) Omit entries in Flat direction from `node`

### LINKED ISSUE #3285: `node` should omit nodes from `Flat` dimensions
I think it'd be cleaner to change `node` so that `Flat` dimensions are omitted.

For example, this would mean that on a grid that is flat in `z` (two-dimensional in `x, y`, we would write initial conditions

```julia
ui(x, y) = # initial u
set!(model, u=ui)
```

Basically, grids that are `Flat` in `z` cannot depend on `z`. So it would be clearer if initial conditions also did not depend on `z`.

### PR DESCRIPTION
**Pull request by glwagner**
Closes #3285

This is a big API change so we'll have to bump a minor version.

This PR changes the behavior of the internal function `node`, which returns a tuple of coordinates at index `i, j, k` and location `lx, ly, lz`, to omit coordinates from the tuple along `Flat` directions. Previously, `node` omitted coordinates only for `Nothing` locations. Now, `Flat` directions are treated similarly to `Nothing` locations.

This change changes the API for using `set!` with functions, and therefore for setting initial conditions, building immersed boundaries, using boundary condition functions, and background fields.

Before this PR, one might write

```julia
grid = RectilinearGrid(size=64, z=(0, 1), topology=(Flat, Flat, Grid))
c = CenterField(grid)
random_numbers(x, y, z) = rand()
set!(c, random_numbers)
```

After this PR, this code has to change because the `x, y` directions are `Flat`. Thus after this PR we will write

```julia
grid = RectilinearGrid(size=64, z=(0, 1), topology=(Flat, Flat, Grid))
c = CenterField(grid)
random_numbers(z) = rand()
set!(c, random_numbers)
```

As a rule, coordinates within functions should never be provided for `Flat` directions or for `Nothing` locations.

There's more discussion about the pros and cons of this choice on #3285.

### CODE REVIEW

**Comment by glwagner** [comment]
Should we also try things like

```julia
node(i, j, k, grid, ::Nothing, ℓy, ℓz) = _node(i, j, k, grid, nothing, , ℓy, ℓz)[1:2]
```

I think we determined there could be a tiny performance loss but it would make the code a little simpler and also easier to read since we don't have to define every combination of locations for `_node`.

I still want to clean up and streamline `xspacings` and `xnodes` (etc) but that'll have to wait for another PR I think.

**Comment by glwagner** [comment]
> Should we also try things like
> 
> ```julia
> node(i, j, k, grid, ::Nothing, ℓy, ℓz) = _node(i, j, k, grid, nothing, , ℓy, ℓz)[1:2]
> ```
> 
> I think we determined there could be a tiny performance loss but it would make the code a little simpler and also easier to read since we don't have to define every combination of locations for `_node`.

Hmm this is more annoying than I thought so I'll leave it for later.

**Comment by glwagner** [comment]
Ok, we now emit warnings like

```julia
julia> include("one_dimensional_diffusion.jl")
┌ Warning: An error was encountered within set! while setting the field
│
│     1×1×128 Field{Center, Center, Center} on RectilinearGrid on CPU
│
│ Note that to use set!(field, func::Function) on a field at location (Center, Center, Center)
│ and on a grid with topology (Flat, Flat, Bounded), func must be callable via
│
│      func(z)
└ @ Oceananigans.Fields ~/Projects/Oceananigans.jl/src/Fields/set!.jl:59
```

File: src/Grids/orthogonal_spherical_shell_grid.jl [hunk 1]
```diff
@@ -1171,6 +1159,14 @@ end
 @inline xnode(i, j, k, grid::OSSG, ℓx, ℓy, ℓz) = xnode(i, j, grid, ℓx, ℓy)
 @inline ynode(i, j, k, grid::OSSG, ℓx, ℓy, ℓz) = ynode(i, j, grid, ℓx, ℓy)
 
+# Definitions for node
+@inline ξnode(i, j, k, grid::OSSG, ℓx, ℓy, ℓz) = λnode(i, j, grid, ℓx, ℓy)
+@inline ηnode(i, j, k, grid::OSSG, ℓx, ℓy, ℓz) = φnode(i, j, grid, ℓx, ℓy)
+@inline rnode(i, j, k, grid::OSSG, ℓx, ℓy, ℓz) = znode(k, grid, ℓz)
+
+ξname(::OSSG) = :λ
+ηname(::OSSG) = :η
```
Comment (navidcy) [inline src/Grids/orthogonal_spherical_shell_grid.jl:1168]: should this be `:φ`?

File: src/Grids/latitude_longitude_grid.jl [hunk 2]
```diff
@@ -574,7 +574,33 @@ coordinates(::LatitudeLongitudeGrid) = (:λᶠᵃᵃ, :λᶜᵃᵃ, :φᵃᶠᵃ
 ##### Grid nodes
 #####
 
-function nodes(grid::LatitudeLongitudeGrid, ℓx, ℓy, ℓz; reshape=false, with_halos=false)
+ξname(::LLG) = :λ
+ηname(::LLG) = :η
```
Comment (navidcy) [inline src/Grids/latitude_longitude_grid.jl:578]:
```suggestion
ηname(::LLG) = :φ
```
?

**Review by navidcy — APPROVED** [review]
noice!

File: src/Fields/set!.jl [hunk 3]
```diff
@@ -19,16 +19,51 @@ function set!(u::Field, v)
...
+        loc_str = tuple_string(location(u))
+        topo_str = tuple_string(topology(u.grid))
+
+        msg = string("An error was encountered within set! while setting the field", '\n', '\n',
+                     "    ", prettysummary(u), '\n', '\n',
+                     "Note that to use set!(field, func::Function) on a field at location ",
+                     "(", loc_str, ")", '\n',
+                     "and on a grid with topology (", topo_str, "), func must be ",
+                     "callable via", '\n', '\n',
+                     "     func(", arg_str, ")", '\n')
```
Comment (tomchor) [inline src/Fields/set!.jl:59]: :heart:

**Comment by tomchor** [comment]
Since this is such a major change in the interface, I think we should get the docs back up again before registering 0.90 (i.e. fix https://github.com/CliMA/Oceananigans.jl/issues/3191).

**Comment by navidcy** [comment]
We don’t know how to do that unfortunately

**Comment by glwagner** [comment]
Happy to hold off registering 0.90.0.

**Comment by tomchor** [comment]
> That’s a good point. Unfortunately, we don’t know how to do that unfortunately. We tried bunch of things. I’ll give it a go.

Can't we simply revert the commit/PR that introduced the issue?

**Comment by glwagner** [comment]
It's not due to a commit (as far as we can tell). It's because we changed a buildkite setting.

**Comment by navidcy** [comment]
It bothers me this issue... I'll try to get to the bottom of it.

**Comment by navidcy** [comment]
I changed a setting. I don't know how to test it though other than registering a new version... But what if it doesn't work?

**Comment by glwagner** [comment]
I think it's not the end of the world if we go a few days without docs. It's a risk but it's not worth losing sleep and mental health over this

**Comment by navidcy** [comment]
Shall I register and see?

**Comment by navidcy** [comment]
I'm doing it.

**Pull request merged**